pact/brackets-review.pact

```diff
@@ -0,0 +1,504 @@
+   (defconst WINNER_PAID:string "winner-paid")
+   ;team has no player assigned
+   (defconst UNASSIGNED:string "unassigned")
+
+  (defschema empty-bracket
+    ;will need to refactor admin to take in a keyset
+    ;not sure how to do this from the frontend though...
+    admin:string
+    bracket:list
+    status:string
+    moneyPool:integer
+    players:[string]
+    winner:string
+    entryFee:integer
+    usernames:[string])
+    ;guard:guard
+
+  (deftable empty-bracket-table:{empty-bracket})
+
+  (defschema bracket-betting
+    ;will need to refactor admin to take in a keyset
+    ;not sure how to do this from the frontend though...
+    admin:string
+    bracket:list
+    status:string
+    moneyPool:integer
+    players:[string]
+    players-bets:list
+    winner:string
+    entryFee:integer)
+
+  (deftable bracket-betting-table:{bracket-betting})
+
+  (defschema users
+    username:string
+    user-guard:guard
+    balance:integer
+    bb-games:list
+    bb-admins:list
+    eb-games:list
+    eb-admins:list
+    games-won:list)
+
+  (deftable users-table:{users})
+
+  (defcap BRACKET-ADMIN-BB (admin-key:string bracket-name:string)
+    (with-read bracket-betting-table bracket-name { "admin":= admin-key-db }
+      (enforce (= admin-key admin-key-db) "you are not the bracket admin")
+    )
+  )
+
+  (defcap BRACKET-ADMIN-EB (admin-key:string bracket-name:string)
+    (with-read empty-bracket-table bracket-name { "admin":= admin-key-db }
+      (enforce (= admin-key admin-key-db) "you are not the bracket admin")
+    )
+  )
+
+  (defcap USER-CAN-TRANSFER (from:string amount:integer)
+    (with-read users-table from { "balance" := from-bal }
+      (enforce (>= from-bal amount) "Insufficient Funds")
+    )
+  )
+
+  (defcap EB-CONTRACT-CAN-TRANSFER (bracket-name:string amount:integer)
+    (with-read empty-bracket-table bracket-name { "moneyPool" := from-bal }
+      (enforce (>= from-bal amount) "Insufficient Funds")
+    )
+  )
+
+  (defcap BB-CONTRACT-CAN-TRANSFER (bracket-name:string amount:integer)
+    (with-read bracket-betting-table bracket-name { "moneyPool" := from-bal }
+      (enforce (>= from-bal amount) "Insufficient Funds")
+    )
+  )
+
+  (defcap IS-REGISTERED-USER (user-key:string)
+    (let ((users (keys users-table)))
+          ;should be user-id
+        (enforce (contains user-key users) "you are not a registered user")
+    )
+  )
+
+  ;needs to be passed as arg to function (the guard)
+  ;user-key is both id and documents user from payload
+  ;usually its keyset in the arguments
+  (defun init-user (user-key:string username:string)
+    ;user id
+    (insert users-table user-key {
+      "username": username,
+      "balance": 100,
+      "user-guard": (read-keyset user-key),
+      "bb-games": [],
+      "bb-admins": [],
+      "eb-games": [],
+      "eb-admins": [],
+      "games-won": []
+    })
+  )
+
+  (defun get-user-info (user-key:string)
+    (with-capability (IS-REGISTERED-USER user-key)
```

Comment:
This test is unnecessary, just let `with-read` fail on a bad key and handle in the front end. Also you might consider just using `read` here, either with a column list, or just return the whole record and filter on the front end. `with-read` is pretty unweildy with lots of variables.